_posts/2021-02-04-justctf21-rusty.md

```diff
@@ -0,0 +1,219 @@
+I didn't manage solve it, so after getting a hint that said that the real flag was hidden elsewhere
+in the `exe`, I had to return to crack this open.
+
+---
+
+### Meet the DOS stub
+
+The Windows Portable Executable format (PE) is designed to be backwards compatible - that is, if a user
+tries to execute a Windows binary under DOS, it will normally just tell the user:
+
+> This program cannot be run in DOS Mode.
+
+This message is not output from the OS; it is generated from a small 16-bit DOS executable section in the
+start of the .exe that is known as the _DOS stub_.
+
+_Normally_ this stub contains the message above and basically `MOV AL, 9 / INT 0x21` which is the DOS equivalent syscall for writing to a File Descriptor, in this case STDOUT.
+
+But when I inspected the header fields in the program, I noticed that the header (Ghidra marks this section as `e_program`) was 4368 bytes long!
+
+#### Extracting e_program
+
+The easiest way (using partially Ghidra) to extract the e_program header section into a separate binary,
+was to select the whole region, and use _Copy Special_ -> _Byte String_ on the context menu, paste it
+into a temporary file, or directly into something like CyberChef, choose the _From Hex_ recipe, and
+download the decoded bin. Of course it is possible to cut out the relevant bytes using command line tools,
+but this time I wanted to check if there was a feasible way of doing it with Ghidra. It would be nice if
+Ghidra had a way to export a section directly as raw, verbatim bytes, or even better, re-import it into
+the project as a separate file.
+
+#### Analyzing the DOS stub
+
+Since I cut out the header region without any PE/MZ header, some non-conventional steps are needed when
+importing it into Ghidra.
+
+![Import raw binary 1](/assets/images/posts/justctf21_rusty/rusty-import-dos-stub.bin-1.png "Import raw binary in Ghidra #1")
+
+![Import raw binary 2](/assets/images/posts/justctf21_rusty/rusty-import-dos-stub.bin-2.png "Import raw binary in Ghidra #2")
+
+To find out where the first instruction is, we need to go back and inspect the MZ header.
+
+`e_ip` (Initial IP value) is set to `0x8a` - but what does it offset of?
+
+Two DOS terms we need to know about:
+
+* A _paragraph_ is `16` bytes (`0x10`)
+* A _page_ is `512` bytes (`0x200`)
+
+Turns out that we need to take page alignment into account. Since the header up to the `e_program`
+section is `0x40` bytes, the current page will be padded with `0x200 - 0x40 = 0x1c0 (448)` bytes
+before the page(s) for the working set of the executable.
+
+![Ghidra byte listing](/assets/images/posts/justctf21_rusty/rusty-dos-stub-bin-listing-1.png "Ghidra raw listing of rusty DOS stub")
+
+At `0x1c0`, we discover the bytes `This program cannot be run in DOS Mode.$`. Notice the
+`$` at the end? Strings handled by the syscalls of `INT 0x21` terminated by a dollar - not by `0`
```

Comment:
Strings handled by the syscalls of `INT 0x21` terminated by a dollar

Kanskje "are terminated by a dollar"?

Reply:
fikset